### ISSUE TITLE
logging every step for 175B model is roughly costing 1 second per step and not getting counted in WPS and UPS

### ISSUE THREAD

**Issue by ngoyal2707** [facebookresearch/metaseq#327]
most likely because we loop over bunch of tensors (gradients / activations norms etc) and move them to cpu for logging.

Weirdly this happens outside of WPS and UPS counters, so we were not noticing this.

**Comment by ngoyal2707** [comment]
How I identified the issue:

added record_function for logging_stats which is after the train_step:

https://github.com/facebookresearch/metaseq/blob/11bf89f3aa128acc44de359aa1de02c275e54f8f/metaseq_cli/train.py#L283-L295


the profile roughly looks something like:
![Screen Shot 2022-09-01 at 4 26 31 PM](https://user-images.githubusercontent.com/7836935/188005874-7cf71aed-9a54-4aec-98bb-bba057601d67.png)


The reason its not recorded in the WPS and UPS is because we log the WPS and UPS at the end of `trainer.train_step` here: https://github.com/facebookresearch/metaseq/blob/11bf89f3aa128acc44de359aa1de02c275e54f8f/metaseq/trainer.py#L831

**Comment by punitkoura** [comment]
Made these changes in meters.py

```diff
diff --git a/metaseq/logging/meters.py b/metaseq/logging/meters.py
index a392fe1..037ec98 100644
--- a/metaseq/logging/meters.py
+++ b/metaseq/logging/meters.py
@@ -7,6 +7,7 @@ import bisect
 import time
 from collections import OrderedDict
 from typing import Dict, Optional
+import logging
 
 
 try:
@@ -45,6 +46,12 @@ class Meter(object):
 
     def reset(self):
         raise NotImplementedError
+    
+    def __str__(self):
+        return str(self.state_dict())
+    
+    def __repr__(self):
+        return str(self.state_dict())
 
     @property
     def smoothed_value(self) -> float:
@@ -259,19 +266,23 @@ class MetersDict(OrderedDict):
         """Get a single smoothed value."""
         meter = self[key]
         if isinstance(meter, MetersDict._DerivedMeter):
-            return meter.fn(self)
+            with torch.autograd.profiler.record_function(f"get_smoothed_values_{key}"):
+                return meter.fn(self)
         else:
-            return meter.smoothed_value
+            with torch.autograd.profiler.record_function(f"get_smoothed_values_{key}"):
+                return meter.smoothed_value
 
     def get_smoothed_values(self) -> Dict[str, float]:
         """Get all smoothed values."""
-        return OrderedDict(
-            [
-                (key, self.get_smoothed_value(key))
-                for key in self.keys()
-                if not key.startswith("_")
-            ]
-        )
+        # logging.info(f"++++++++++++++++++++++++ Inside meters::get_smoothed_values - {self.keys()} - {self.values()} +++++++++++++++++++++++++")
+        with torch.autograd.profiler.record_function("get_smoothed_values_inner_inner"):
+            return OrderedDict(
+                [
+                    (key, self.get_smoothed_value(key))
+                    for key in self.keys()
+                    if not key.startswith("_")
+                ]
+            )
 
     def reset(self):
         """Reset Meter instances."""
(fairseq-20220503) punitkoura@fairwus3-1-login-1:~/src/metaseq$ 
```

To look at the tensors being transferred from GPU to CPU, I printed out the self.values (see the commented out logging.info statement)

```
Inside meters::get_smoothed_values - odict_keys(['actv_norm', 'pos_norm', 'tok_norm', 'emb_norm', 'docsperex', 'loss', 'ppl', 'wps', 'ups', 'wpb', 'bsz', 'num_updates', 'lr', 'gnorm', 'clip', 'loss_scale', 'scale_window', 'train_wall', 'cuda_gb_allocated', 'cuda_gb_reserved', 'cuda_gb_free']) - odict_values(
    
    [
        {'val': tensor(0.9050, device='cuda:0', dtype=torch.float64), 'sum': tensor(0.9050, device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 3}
        {'val': tensor(0.1663, device='cuda:0', dtype=torch.float64), 'sum': tensor(0.1663, device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 3}
        {'val': tensor(0.2350, device='cuda:0', dtype=torch.float64), 'sum': tensor(0.2350, device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 3}
        {'val': tensor(0.0009, device='cuda:0', dtype=torch.float64), 'sum': tensor(0.0009, device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 3}
        {'val': tensor(6.4805, device='cuda:0', dtype=torch.float64), 'sum': tensor(6.4805, device='cuda:0', dtype=torch.float64), 'count': 1, 'round': None}
        {'val': tensor(15.6360, device='cuda:0', dtype=torch.float64), 'sum': tensor(8197783.0205, device='cuda:0', dtype=torch.float64), 'count': tensor(524288., device='cuda:0', dtype=torch.float64), 'round': 3}
        {}
        {'init': 0.02124020690098405, 'n': 0, 'round': 1},
        {'init': 0.021953568793833256, 'n': 0, 'round': 2}
        {'val': tensor(524288., device='cuda:0', dtype=torch.float64), 'sum': tensor(524288., device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 1}
        {'val': tensor(256., device='cuda:0', dtype=torch.float64), 'sum': tensor(256., device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 1}
        {'val': 1, 'sum': 0, 'count': 0, 'round': None},
        {'val': 8.39160839160839e-07, 'sum': 0, 'count': 0, 'round': None}
        {'val': tensor(12.0024, device='cuda:0', dtype=torch.float64), 'sum': tensor(12.0024, device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 3}
        {'val': tensor(100., device='cuda:0', dtype=torch.float64), 'sum': tensor(100., device='cuda:0', dtype=torch.float64), 'count': 1, 'round': 1}
        {'val': 2.0, 'sum': 0, 'count': 0, 'round': 4}
        {'val': 128, 'sum': 0, 'count': 0, 'round': 4}
        {'sum': 4.089079717174172, 'n': 0.0, 'round': 0}
        {'val': 5.20842170715332, 'sum': 0, 'count': 0, 'round': 1}
        {'val': 9.400390625, 'sum': 0, 'count': 0, 'round': 1}
        {'val': 73.96125602722168, 'sum': 0, 'count': 0, 'round': 1}
    ]
)
```

We can see that the number of tensors is fixed for most keys (expect for ppl). Will look into ppl as well. It's a bit different since it is a derived meter (metaseq.logging.meters.MetersDict._DerivedMeter)
